Installation issue: gcc

<!-- Thanks for taking the time to report this build failure. To proceed with the report please:

1. Title the issue "Installation issue: <name-of-the-package>".
2. Provide the information required below.

We encourage you to try, as much as possible, to reduce your problem to the minimal example that still reproduces the issue. That would help us a lot in fixing it quickly and effectively! -->

### Steps to reproduce the issue

<!-- Fill in the exact spec you are trying to build and the relevant part of the error message -->
```console
$ spack install gcc @10.3.0 +graphite %apple-clang @12.0.5
```

This builds GCC with the Apple Clang compiler shipped with macOS 11.3. The previous version was 12.0.0, which worked fine for building GCC.

### Information on your system

<!-- Please include the output of `spack debug report` -->

```
$ spack debug report
* **Spack:** 0.16.1-2404-303e506f2b
* **Python:** 3.8.5
* **Platform:** darwin-bigsur-skylake
* **Concretizer:** original
```

<!-- If you have any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.) you can add that here as well. -->

### Additional information

<!-- Please upload the following files. They should be present in the stage directory of the failing build. Also upload any config.log or similar file if one exists. -->
* [spack-build-out.txt.gz](https://github.com/spack/spack/files/6387605/spack-build-out.txt.gz)
* [spack-build-env.txt](https://github.com/spack/spack/files/6387598/spack-build-env.txt)

<!-- Some packages have maintainers who have volunteered to debug build failures. Run `spack maintainers <name-of-the-package>` and @mention them here if they exist. -->

@michaelkuhn

### General information

<!-- These boxes can be checked by replacing [ ] with [x] or by clicking them after submitting the issue. -->
- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

I spent an hour at least debugging this morning... this happened with the same "apple-clang" version that previously worked, but I guess the rest of the toolchain (as, nm, etc.) changed. I wish there was a way to encode that into the DAG... maybe whenever the compilers-as-packages thing happens?

Error message:
```
gcc/alias.o differs
gcc/analyzer/analyzer-logging.o differs
...
libiberty/xatexit.o differs
make[2]: *** [compare] Error 1
make[1]: *** [stage3-bubble] Error 2
make: *** [all] Error 2
```

The simplest file change was `libiberty/lbasename`. Here's the diff of `stage{2,3}-libiberty/lbasename.o` piped through `strip -S` and disassembled with `objdump -D`:
```
$ diff -u stage2-nodebug.s stage3-nodebug.s
--- stage2-nodebug.s	2022-02-05 13:28:16.683152063 -0500
+++ stage3-nodebug.s	2022-02-05 13:28:20.223081399 -0500
@@ -1,5 +1,5 @@

-stage2.o:	file format mach-o 64-bit x86-64
+stage3.o:	file format mach-o 64-bit x86-64


 Disassembly of section __TEXT,__text:
@@ -58,8 +58,8 @@
       90: 0f b6 07                     	movzbl	(%rdi), %eax
       93: 49 89 f8                     	movq	%rdi, %r8
       96: 84 c0                        	testb	%al, %al
-      98: 0f 84 13 00 00 00            	je	0xb1 <_lbasename+0x21>
-      9e: 66 90                        	nop
+      98: 74 17                        	je	0xb1 <_lbasename+0x21>
+      9a: 66 0f 1f 44 00 00            	nopw	(%rax,%rax)
       a0: 48 83 c7 01                  	addq	$1, %rdi
       a4: 3c 2f                        	cmpb	$47, %al
       a6: 4c 0f 44 c7                  	cmoveq	%rdi, %r8
```
the function definition (lbasename) is identical to the `lbasename_unix` above, but [an optimization causes the assembler to change the code layout](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100340) by changing [one kind of noop to another](http://john.freml.in/amd64-nopl).

I'm going to push a fix that adds a conflict to force bootstrapping off for this version of xcode.

--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -226,8 +226,14 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     # Binutils can't build ld on macOS
     conflicts('+binutils', when='platform=darwin')
 
+    # Bootstrap comparison failure:
+    #   see https://github.com/spack/spack/issues/23296
+    #   https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100340
+    #   on XCode 12.5
+    conflicts('+bootstrap', when='@:10 %apple-clang@12.0.5')
+
     # aarch64/M1 is supported in GCC 12+
-    conflicts('@:11.99', when='target=aarch64: platform=darwin',
+    conflicts('@:11', when='target=aarch64: platform=darwin',
               msg='Only GCC 12 and newer support macOS M1 (aarch64)')
 
     # Newer binutils than RHEL's is required to run `as` on some instructions

--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -252,7 +252,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
             # https://github.com/Homebrew/homebrew-core/pull/39041
             patch('https://raw.githubusercontent.com/Homebrew/formula-patches/master/gcc/8.3.0-xcode-bug-_Atomic-fix.patch',
                   sha256='33ee92bf678586357ee8ab9d2faddf807e671ad37b97afdd102d5d153d03ca84',
-                  when='@6:8')
+                  when='@6:8.3')
         if macos_version() >= Version('10.15'):
             # Fix system headers for Catalina SDK
             # (otherwise __OSX_AVAILABLE_STARTING ends up undefined)

--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -230,7 +230,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     #   see https://github.com/spack/spack/issues/23296
     #   https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100340
     #   on XCode 12.5
-    conflicts('+bootstrap', when='@:10 %apple-clang@12.0.5')
+    conflicts('+bootstrap', when='@:11.1 %apple-clang@12.0.5')
 
     # aarch64/M1 is supported in GCC 12+
     conflicts('@:11', when='target=aarch64: platform=darwin',

